### PR TITLE
LPS-196350 log error when checking for a feature flag that is not present in the specified scope

### PR DESCRIPTION
**Pull request by thiago-buarqque**
[LPS-196350](https://liferay.atlassian.net/browse/LPS-196350)

Goal
=======
Ensure that developers understand the specific scope of the feature flag they are checking when using `FeatureFlagManagerUtil.isEnabled("LPS-xxx")`. If it's not possible to find the feature flag, they should pass the `companyId` or set it in the `CompanyThreadLocal`. It's important to avoid checking for an instance-scoped feature flag in the system-level FeatureFlagsBag, and vice versa. This helps prevent confusion and ensures accurate evaluation of feature flags within the intended scope.

Proposed Solution
========
1. Log an error message warning that the Feature Flag is not present in the specified scope.
2. Add to the [docs](https://liferay.atlassian.net/wiki/spaces/ENGR/pages/1604714779/Feature+Flags) how this error occurs and how to prevent it. For instance, the text to be added to the docs is:

> If you're checking for a Feature Flag status, but companyId is not available either where you're checking the Feature Flag or in CompanyThreadLocal, you should consider refactoring the check to where you have the companyId or make the Feature Flag system scoped in case you can't refactor the code (refer to the System Feature Flags for more details).

Feature flags I catch along the way
========
- LPS-183661:
    - Occurred at: `ElasticsearchIndexWriter.java:407`, `CompanyIndexFactory.java:63` and `CompanyIndexFactoryHelper.java:98`) 
    - Approach: [recently removed by Yu](https://github.com/liferay/liferay-portal/commit/25c769ba6f6598491fd14065c2f03d69aaed77cd)
- COMMERCE-9410 
    - Occurred at: `CommerceOrderInfoItemCapabilitiesProvider.java:30`,  
    - Approach: [@Danny Situ](https://liferay.slack.com/team/UFN5ER96X) said it makes sense to be turned into system scoped
- LPS-180198 
    - Occurred at: `MembershipsUserFieldExpressionHandlerRegistrator.java`
    - Approach: [@Stian Sigvartsen](https://liferay.slack.com/team/U04TD8UQ3) said it was okay to be turned into system scoped
- LPS-180328 
    - Approach: passed companyId `LayoutImpl.java:1521`
- LPS-157670 
    - Approach: is used on `StartupHelperUtil:121`  to upgrade the portal, so it's system scoped
- LPS-184255 
    - Approach: passed companyId `com.liferay.journal.web.internal.model.listener.DDMStructureModelListener.java:41`
- LRAC-10632
    - Occurred at: `BaseModelListener.java:56`
    - Approach: [@Marcellus Tavares](https://liferay.slack.com/team/U04CWP1ER) said it should be system scoped
- LPS-195205 
    - Approach: passed companyId in `LayoutLocalServiceImpl.java:239`
- LPS-174816 
    - Approach: passed companyId in `StoreAreaAwareStoreWrapper.java:222`
- LPS-122920 
    - Approach: passed companyId in `BaseTextEmbeddingModelDocumentContributor.java:153`
- LPS-180328
    - Approach: passed companyId in `com.liferay.layout.content.page.editor.web.internal.model.listener.LayoutModelListener:28`

### CODE REVIEW

**Comment by thiago-buarqque** [comment]
Unrelated SF errors. Waiting for fix on master since who removed the Feature Flag must remove it from portal.properties and make sure there are no other references (currently there are references in JSP [[see example](https://github.com/liferay-peds/liferay-portal/blob/2f31084975b8217be699b430883a9f0a97c1eefc/modules/dxp/apps/search-experiences/search-experiences-web/src/main/resources/META-INF/resources/sxp_blueprint_admin/edit_sxp_blueprint.jsp#L38)]) and I guess SF doesn't catch those.

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
@eltonccdantas I'm waiting for the CI results to see what I can do

**Comment by thiago-buarqque** [comment]
Fixing the CI errors, they were expected

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
Running CI again, too many tests to run manually. Now I'm logging the stack trace to find the source of each errors and contact the teams regarding them.

**CI is expected to fail**

**Comment by thiago-buarqque** [comment]
I got the source for the feature flags errors, I'm contacting the teams regarding changing them to system scoped (if possible)

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:reevaluate:1397401_7631

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
Rebasing since the last build failed on downstream failed and I couldn't check the logs for each test

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
Hey @ethib137. Yeah, I had checked with the teams. Once I finish all feature flags I'll describe each of them here, what was the fix, who approved and also ping who approved in this PR. For now I'm running CI again since it seems that a new feature flag was introduced and I want to know where the companyId is not being properly set.

**Comment by thiago-buarqque** [comment]
Contacting the search team regarding a new FF that just appeared in the error logs...

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

File: modules/dxp/apps/saml/saml-opensaml-integration/src/main/java/com/liferay/saml/opensaml/integration/internal/field/expression/handler/MembershipsUserFieldExpressionHandlerRegistrator.java [hunk 1]
```diff
@@ -27,7 +28,7 @@ public class MembershipsUserFieldExpressionHandlerRegistrator {
 
 	@Activate
 	protected void activate(BundleContext bundleContext) {
-		if (!_featureFlagManager.isEnabled("LPS-180198")) {
+		if (!FeatureFlagManagerUtil.isEnabled("LPS-180198")) {
```
Comment (drewbrokke) [inline modules/dxp/apps/saml/saml-opensaml-integration/src/main/java/com/liferay/saml/opensaml/integration/internal/field/expression/handler/MembershipsUserFieldExpressionHandlerRegistrator.java:31]: Is the `_featureFlagManager` field now unused?

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
@ethib137 @drewbrokke check the description and the changes I made. Let me know if you have any questions.

**Review by ethib137 — COMMENTED** [review]
Looks good, I just added one question inline.

File: portal-impl/src/com/liferay/portal/service/impl/LayoutLocalServiceImpl.java [hunk 2]
```diff
@@ -236,7 +236,8 @@ public Layout addLayout(
 
 		if (system &&
 			((Objects.equals(type, LayoutConstants.TYPE_ASSET_DISPLAY) &&
-			  !FeatureFlagManagerUtil.isEnabled("LPS-195205")) ||
+			  !FeatureFlagManagerUtil.isEnabled(
+				  user.getCompanyId(), "LPS-195205")) ||
```
[the remark this answers is not in the document]
Comment (drewbrokke) [reply portal-impl/src/com/liferay/portal/service/impl/LayoutLocalServiceImpl.java:240]: @ethib137 the `userId` of this user is passed into the method, so IMO it should be safe to assume that the companyId will match. Super admins cannot log into other virtual instances even though they may have board permissions, so a mismatch is very unlikely. It is actually a very common pattern to derive the `companyId` from the passed in `userId` in a service's create method, so if there's a problem here, it won't be a new one.

[the remark this answers is not in the document]
Comment (thiago-buarqque) [reply portal-impl/src/com/liferay/portal/service/impl/LayoutLocalServiceImpl.java:240]: @drewbrokke Thanks Drew

**Comment by thiago-buarqque** [comment]
The last CI run was last week, I updated the branch and I'm running CI again to make sure there is no new feature flag anywhere. After that I ask for a quick decision on whether send this to Brian so (is common to have new FFs)

**Comment by thiago-buarqque** [comment]
ci:test:relevant

File: portal-impl/src/com/liferay/portal/service/impl/LayoutLocalServiceImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (drewbrokke) [reply portal-impl/src/com/liferay/portal/service/impl/LayoutLocalServiceImpl.java:240]: Anytime 👍

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
Hey @eltonccdantas , the errors doesn't seem related to my changes, could help me out here?

**Comment by eltonccdantas** [comment]
Let me take a look 🕵🏻

**Comment by eltonccdantas** [comment]
Hey @thiago-buarqque , lots of `testRun: Failed to run test on CI`, lets trigger another run

**Comment by thiago-buarqque** [comment]
@eltonccdantas  Okay

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
Hey @thiago-buarqque, I reckon this pull request has some unique failures on these tests:


1. [com.liferay.search.experiences.internal.blueprint.test.SXPBlueprintSearchResultTest](https://testray.liferay.com/home/-/testray/case_results/2441133675)
2. [com.liferay.layout.search.test.LayoutMultiLanguageSearchTest](https://testray.liferay.com/home/-/testray/case_results/2441132292)
3. [com.liferay.layout.search.test.LayoutIndexerIndexedFieldsTest](https://testray.liferay.com/home/-/testray/case_results/2441132495)


Would you take a closer look on their failure messages? 

[Comparison run](https://testray.liferay.com/home/-/testray/runs/compare?formDate=1697457783736&testrayRunIdA=2440865274&testrayRunIdB=2441128346)

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
Two new feature flags were added, logging cause of error to fix them.

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
@drewbrokke [This](https://github.com/liferay/liferay-portal/commit/bc8ff3c40e28d949434bbd39684283662d245be0#diff-2aec3ebeb96cbc33583c163550d9948888a3d9e73decc664d825b9f51d4e681eR207) change was introduced yesterday and it checks for feature flags in a generic way. [I retrieved the companyId from the BatchEngineUnitConfiguration](https://github.com/liferay-peds/liferay-portal/pull/277/commits/26e99e7c1be112cd0db8bba24c64a48d10523e39) and passed it for the feature flag check. I'm highlighting this because the check is generic.

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
> @drewbrokke [This](https://github.com/liferay/liferay-portal/commit/bc8ff3c40e28d949434bbd39684283662d245be0#diff-2aec3ebeb96cbc33583c163550d9948888a3d9e73decc664d825b9f51d4e681eR207) change was introduced yesterday and it checks for feature flags in a generic way. [I retrieved the companyId from the BatchEngineUnitConfiguration](https://github.com/liferay-peds/liferay-portal/pull/277/commits/26e99e7c1be112cd0db8bba24c64a48d10523e39) and passed it for the feature flag check. I'm highlighting this because the check is generic.

Passing the companyId doesn't work, the companyId is not set to the BatchEngineUnitConfiguration. I'll contact the owner of these changes and ask how things work and see how I can proceed

**Comment by thiago-buarqque** [comment]
I opened [this](https://liferay.slack.com/archives/C5E1CRLJY/p1697717425376029) thread to discuss things further on. Changes were sent and now it's supposed to work. I'm rebasing and running CI

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
**Closing PR**

I opened this PR since this one has too many CI comments due to changes introduced by other teams (new FFs) and other CI fails (like finishing and not compiling the results).

I'll pass all the new comments since last review to the new PR.

Please @drewbrokke and @ethib137 , refer to #1234556 now

**Comment by thiago-buarqque** [comment]
The [last](https://testray.liferay.com/home/-/testray/runs?testrayBuildId=2444472283) CI run failed to unrelated changes.

**Pull request closed without merging**